ENH:MAINT:special:Rewrite amos F77 code

ilayn opened the pull request:
#### Reference issue
Part of #18566 

Relevant issues will be added 

#### What does this implement/fix?
The rewrite of the AMOS library. 

#### Additional information
Not as big as cdflib, should be tested meticulously for the edge cases.

Total gotos left: ~40 (out of too many)


- [x] Add missing documentation from Fortran sources
- [x] Add tests
- [x] Get some benchmarks

ilayn commented:
I managed to make some progress in a quite unexpected direction. 

Because the code is so unorthodox, doing real arithmetic with complex ops, jumping from one for loop into the middle of another and feeling no shame and so on, Cythonization didn't cut it so I switched to C as another goto infested lang. Anyways, I quickly realized why I stopped coding in these langs because it's been roughly 30 years and still nobody managed to agree on what a complex number type should be. So probably at some point `#include <complex.h>` will be a headache for somewhere some system. 

It's been really challenging and I unrolled as many GOTOs as I could. At least we have `airy`/`airye` function in working state. I tested a bit and results agree with fortran code. We are at about 25% and I'll clean the cython code when I'm done 😢 

 If you have objections to style or strategy please tell me as I hardly know what I am doing.

ilayn commented:
I am pretty confident about `airy` and `biry` code in C as I tested in many regions and compared with Fortran code and WolframAlpha. 

I'll test the the remaining Bessel ones and start dealing with the wrappers.

ilayn commented:
Tested on Win10 and Linux Mint. Only one test is barely failing rel err `1.09e-13` instead of `1.0e-13`. Let's see what happens in the CI.

Happy new year all 🎉 🎆

ilayn commented:
Status update; 

- Py3.12 job does not understand the `.real` and `.imag` attributes of the NumPy Complex doubles (See the failing job). Is this anything related to NumPy 2.0? 
- I tried to find the source of the mismatch in the test 
![image](https://github.com/scipy/scipy/assets/1303842/c36f46bd-c27c-45a0-b846-47e31c042c32)

    but whatever is happening is happening in the wrappers probably at the lowest registers between C and Fortran. Unfortunately                          this ~1e-58 number is then multiplied with a very large number and the result is not possible to hit that same relative accuracy. Hence I made the test relative tolerance to be `2e-13` instead of `1e-13`. 

If the first one is resolved and second item is acceptable, then this is ready to be reviewed.

lucascolley commented:
> Py3.12 job does not understand the .real and .imag attributes of the NumPy Complex doubles (See the failing job). Is this anything related to NumPy 2.0?

I'm just guessing but is https://github.com/numpy/numpy/pull/24085 related?

ilayn commented:
Ah yes, probably. Let me use the `NPY_CREAL` and `NPY_CIMAG` just to be on the safe side.

rgommers commented:
Impressive effort @ilayn!

This isn't quite reviewable line by line, so I guess we should look at test coverage, build warnings, etc. It maybe also be useful to run clang-format over the new code now.

rgommers commented on scipy/special/_amos.c:
```diff
@@ -0,0 +1,6401 @@
+static int amos_unk1(double complex, double, int, int, int, double complex *, double, double, double);
+static int amos_unk2(double complex, double, int, int, int, double complex *, double, double, double);
+static int amos_uoik(double complex, double, int, int, int, double complex *, double, double, double);
+static int amos_wrsk(double complex, double, int, int, double complex *, double complex *, double, double, double);
```
Should these be in the header file instead?

ilayn replied:
Not really a C veteran but if I understand correctly static keeps it local to this file and does not pollute the namespace of the library. Or at least that's my limited understanding of things you wish to keep local. So these are the prototypes for such elements.

lucascolley commented:
> It maybe also be useful to run clang-format over the new code now.

https://github.com/ilayn/scipy/pull/6 does this with the `.clang-format` file in `special/special`, however it seems like there is no option to stop it from adding whitespace around binary operators... probably not a good idea.

ilayn commented:
> This isn't quite reviewable line by line, so I guess we should look at test coverage, build warnings, etc. 

Indeed, that's one of the tricky things to do with such rewrite but our test coverage is pretty ruthless and discovered quite many bugs. Hence while it is possible that some `if` branch is missed, still the majority of the functionality is there.

> It maybe also be useful to run clang-format over the new code now.

Judging by what @lucascolley showed it is a bit too intrusive in the binary operator spacing but I liked a few changes it made. I'll do that semi-manually. It turns out that PEP8 has really conditioned me to write narrow code and C folks are taking none of it :)

ilayn commented:
Reorganized the commits and cleaned up the history, removed the cython parts. So if no additional comments are coming in, I am ready for clicking this in.

j-bowhay commented:
> This isn't quite reviewable line by line, so I guess we should look at test coverage, build warnings, etc. It maybe also be useful to run clang-format over the new code now.

I checked the test coverage with gcov and it is around 70% but some larger branches aren't hit. however, I ran out of time to work out if those are branches we don't ever use (some functions switch between cephes and amos) or missing test coverage

steppi commented:
I'll finish reviewing and hopefully merge the cdflib PR by the end of the week and hope to have this one done next week.

h-vetinari commented:
> Anyways, I quickly realized why I stopped coding in these langs because it's been roughly 30 years and still nobody managed to agree on what a complex number type should be.

We have MSVC to thank for that. 😑

steppi commented on scipy/special/_amos.c:
```diff
@@ -0,0 +1,6206 @@
+    }
+    /* 40 */
+    if ((az < rl) || (az+az >= dfnu*dfnu)) {
+        nw = amos_mlri(z, fnu, kode, n, cy, tol);
```
```suggestion
        //
        // MILLER ALGORITHM NORMALIZED BY THE SERIES
        //
        nw = amos_mlri(z, fnu, kode, n, cy, tol);
```

This is where this comment should go.

steppi reviewed (changes requested):
It was quite a feat to untangle and translate all of this, the Fortran is not at all easy to understand. Bessel functions can be tricky to compute and despite its tortured code style, AMOS is actually pretty solid, so I wanted to give this a very close look.

What I decided to do is pick one of the most difficult functions and study the code in depth, comparing to the original Fortran and Donald Amos' [paper](https://www.osti.gov/servlets/purl/6148453). I chose `amos_binu` which computes the modified Bessel function of the first kind in the right half plane. If this part had been translated perfectly, I would have had more confidence in everything, since it is among the most difficult parts to get right.

I found an incorrect condition which causes degraded performance in some cases. Just the kind of little mistake I could have made too if I was the one who translated this. It makes me think I'll need to look over everything with a similar level of care. I don't quite have bandwidth for that at the moment. It would probably be several days of work.

@ilayn, I don't know what your bandwidth looks like, but you might want to trace through all of the control flow again to check if you made any other little mistakes like that. Also, I found a comment in the wrong place, it'll be helpful if you make sure all of those are where they belong.

 I found it helpful to look through Amos' paper I linked to above. Despite all of the spaghetti, most of what's happening here is that different methods are being used in different regions of $\mathbb{R}\times \mathbb{Z}$. The unstructured Fortran code makes the logic seem much more complicated than it is.

ilayn commented on scipy/special/_amos.c:
```diff
@@ -0,0 +1,6206 @@
+    double complex csgn, cspn, c1, c2, zn, cy[2];
+    double arg, ascle, az, cpn, dfnu, fmr, sgn, spn, yy;
+    int inu, iuf, nn, nw;
+    double pi = 3.14159265358979324;
```
I wanted to do that indeed but then I got some strange bit divergence in some results and decided to keep it like this as I don't know if libc is doing something else. I have no opinions either way. One thing I learned is that not all platforms define these as the implementation standard is kept open, (don't remember where I read it though).

steppi replied:
I see, that makes sense. We should keep it like it is for now in any case. There's always time to rework this in the future.

rgommers replied:
`M_PI` & friends aren't completely portable IIRC, although they are mostly if you define `_USE_MATH_DEFINES` (which is done automatically for all Cython-generated code).

In some other places there's this kind of bullet-proofing:
```C
#ifndef M_PI
#define M_PI 3.14159265358979323846
```

ilayn commented:
Ha, it's not less feat to go through this so thanks @steppi 

Funny that you picked binu because I think I wrote that one at least 3 times. And the part you stumbled upon is exactly where I didn't get right apparently, the conditions are so weirdly tangled that I made a true false diagram for it. I'm surprised that it is still broken. 

So we are looking at this flow 

https://github.com/scipy/scipy/blob/87f8428b227dd8962d8658482bd8a150a31292a4/scipy/special/amos/zbinu.f#L34-L47

So first jumps, then falls through then jumps further and if not then falls through. The reason I have those extra conditions are these if else sillyness. Fortunately we don't have another occurence of this so I am pretty confident about the remaining parts since I tested quite a number of them.

ilayn commented on scipy/special/_amos.c:
```diff
@@ -0,0 +1,6206 @@
+    //
+    if (az >= rl) {
+        if ((dfnu <= 1.0) || ((dfnu > 1.0) && (az+az >= dfnu*dfnu))) {
+            //
+            // MILLER ALGORITHM NORMALIZED BY THE SERIES
+            //
```
This is the goto 70 block. so probably I confused the comments.

ilayn commented on scipy/special/_amos.c:
```diff
@@ -0,0 +1,6206 @@
+        }
+    }
+    /* 40 */
+    if ((az < rl) || (az+az >= dfnu*dfnu)) {
```
Yes but that's included at the top in the 30 conditions and if fails gets out of that if block to here. (Or at least that's what I had in mind). So it's not handled at the 40 block but inside the 30 block. But apparently there is a mistake somewhere.

ilayn commented:
OK apparently I fixed this in my local work but did not pass it over to the PR

![image](https://github.com/scipy/scipy/assets/1303842/fca83ffa-f4d5-43e5-a774-a47dbb635036)

This binu is such an annoying piece of code.

steppi commented:
> This binu is such an annoying piece of code.

Yeah, it's the trickiest one in my opinion. If that had been perfect, I think I would have just gave everything else a spot check and merged. I'm feeling better about things again after your explanations. I should be able to find time to finish reviewing within the coming week.

lucascolley commented on scipy/special/_amos.c:
```diff
@@ -0,0 +1,6206 @@
+    // ASYMPTOTIC EXPANSION FOR LARGE Z
+    //
+    if (az >= rl) {
+        if ((dfnu <= 1.0) || ((dfnu > 1.0) && (az+az >= dfnu*dfnu))) {
```
In which case I believe the suggestion should be 
```suggestion
        if ((dfnu <= 1.0) || (az+az >= dfnu*dfnu)) {
```
(`||` instead of `&&` and one less end-parenthesis)

steppi replied:
Thanks, that's right. Just a typo. Fixed now. What you wrote is what I actually have locally.

ilayn commented:
No worries, that's really good catch. Let's hope I didn't make more labor intensive mistakes.

ilayn commented:
@steppi apologies for getting back to this again but I think you found a bug in the fortran code itself. Because when I check it with the fortran code itself this is what I get 

``` 
BESI 225: BESK CALLING BINU
 BINU  35: GOING TO 40?   21.213203435596427        21.784271729432426     
 BINU  55: DID NOT GO TO 70 FROM 40
 BINU  59: CALLING UOIK   15.000000000000000        15.000000000000000
 UNIK 135: SR    1.0002612190046860        2.2858395497944661E-002
 UNIK 198: CWRK[15]  0.10036965266287727       -1.1466953394589764E-003
 BINU  59: RETURNED UOIK WITH NW           0
 BINU  63: CY[0]   0.0000000000000000        0.0000000000000000
 BINU 113: CALLING BUNI
 BUNI 158: CALLING UNI1
 UNIK 135: SR    1.0002612190046860        2.2858395497944661E-002
 UNIK 198: CWRK[15]  0.10036965266287727       -1.1466953394589764E-003
 UNIK 135: SR    1.0002612190046860        2.2858395497944661E-002
 UNIK 198: CWRK[15]  0.10036965266287727       -1.1466953394589764E-003
 UNIK 178: K AFTER LOOP           9
 UNIK 181: K AFTER LOOP           9
 UNIK 198: SUM  0.99916734689966913        1.1361412899521058E-004
 UNIK 198: PHI   4.0041698116427993E-002  -4.5746525364945902E-004
 UNIK 198: CWRK[15]  0.10036965266287727       -1.1466953394589764E-003
 UNIK 198: CON[0]  0.39894228040143270
 UNI1 125: CSR[           1 ] =   1.0000000000000000
 UNI1 125: S2  -2.0645908843310795E-055  -1.1127773731040390E-055
 UNI1 125: Y[           0 ] =   -2.0645908843310795E-055  -1.1127773731040390E-055
 BINU 113: RETURNED BUNI WITH NW           0
 BINU 113: CY[0]  -2.0645908843310795E-055  -1.1127773731040390E-055
 BINU 113: CY[1]   1.6198761276743044E-312  -2.0645908843310795E-055
 BESI 228: BESK RETURNED FROM BINU WITH NZ =           0
(-2.0645908843310795e-55, -1.112777373104039e-55, 0, 0)
```

This is what you get from the wrong SciPy result above. The reason I am convinced that this is a Fortran bug is because my code also goes into the same branches and not into UOIK as you mentioned it should go.

```
_amos.c:890: BESI CALLING BINU
_amos.c:1334: BINU AZ < RL check 21.2132034355964265 <? 21.7842717294324260
_amos.c:1334: BINU dfnu < RL check 99.2000000000000028 <? 1.0
_amos.c:1334: BINU az+az < dfnu*dfnu check 42.4264068711928530 <? 198.4000000000000057
_amos.c:1373: BINU CALLING UOIK
_amos.c:1373: BINU with z 15.0000000000000000 15.0000000000000000
_amos.c:3665: UNIK sr = 1.000261219004686 0.02285839549794466
_amos.c:3709: UNIK: cwrk[15]  0.1003696526628773 -0.001146695339458976
_amos.c:1373: BINU RETURNED FROM UOIK NW = 0
_amos.c:1373: BINU cy[0] 0 0
_amos.c:1394: BINU CALLING BUNI
_amos.c:2156: BUNI CALLING UNI1
_amos.c:3665: UNIK sr = 1.000261219004686 0.02285839549794466
_amos.c:3709: UNIK: cwrk[15]  0.1003696526628773 -0.001146695339458976
_amos.c:3665: UNIK sr = 1.000261219004686 0.02285839549794466
_amos.c:3709: UNIK: cwrk[15]  0.1003696526628773 -0.001146695339458976
_amos.c:3726 UNIK: K AFTER LOOP 9
_amos.c:3749: UNIK AFTER INIT=0 case INIT = 9
_amos.c:3709: UNIK: sum  0.9991673468996691 0.0001136141289952106
_amos.c:3709: UNIK: phi  0.04004169811642799 -0.000457465253649459
_amos.c:3709: UNIK: cwrk[15]  0.1003696526628773 -0.001146695339458976
_amos.c:3709: UNIK: con[0]  0.3989422804014327
_amos.c:3342: UNI1 csr[1] 1
_amos.c:3342: UNI1 s2 -2.064590884330946e-55 -1.112777373104005e-55
_amos.c:3342: UNI1 y[0] -2.064590884330946e-55 -1.112777373104005e-55
_amos.c:2156: BUNI RETURNED FROM UNI1 NW = 0
_amos.c:1394: BUNI y[0] -2.064590884330946e-55 -1.112777373104005e-55
_amos.c:1394: BINU RETURNED FROM BUNI NW = 0, NLAST 0
_amos.c:1394: BINU cy[0] -2.064590884330946e-55 -1.112777373104005e-55
_amos.c:892: BESI RETURNED FROM BINU WITH NZ = 0
0 0 -2.0645908843309458552e-55 -1.1127773731040049375e-55i
```

I guess copied the Fortran code a bit too good.

ilayn commented:
Hmm I am more confused now because then my fixed version is diverged from the Fortran code itself.

ilayn commented:
OK the current SciPy which uses Fortran also does not pass 1e-15 rtol level too

```python
C:\Users\ilhan> ipython
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.19.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import scipy

In [2]: scipy.__version__
Out[2]: '1.12.0rc1'

In [3]: import scipy.special as ssp

In [4]: ssp.iv(99.2, 15+15j)
Out[4]: (-2.0645908843310788e-55-1.1127773731040388e-55j)

In [11]: assert_allclose(ssp.iv(99.2, 15+15j), mpmath,rtol=1-15)

[..snip..]

AssertionError:
Not equal to tolerance rtol=-14, atol=0

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 1.28411573e-68
Max relative difference: 5.47508521e-14
 x: array(-2.064591e-55-1.112777e-55j)
 y: array(-2.064591e-55-1.112777e-55j)
```


So I think I'll just take a break for today and tomorrow I'll do another round.

steppi commented on scipy/special/_amos.c:
```diff
@@ -0,0 +1,6206 @@
+    /* 60 */
+    if (az <= rl) {
+        /* 70 */
+        nw = amos_mlri(z, fnu, kode, n, cy, tol);
+        nz = -1;
+        if (nw == -2) { nz = -2; }
+        return nz;
+    }
```
I've found that the failing test for `v = -20.0, z = -10.0 + 0j` can be traced to the handling after the result comes out of `amos_mlri` here. I've confirmed that `cy` comes out correctly here after my suggestion is applied, but the incorrect `nz` is returned, causing return with error when control passes back to `amos_besi`.

steppi commented on scipy/special/_amos.c:
```diff
@@ -0,0 +1,6206 @@
+    if (az <= rl) {
+        /* 70 */
+        nw = amos_mlri(z, fnu, kode, n, cy, tol);
+        nz = -1;
+        if (nw == -2) { nz = -2; }
+        return nz;
```
```suggestion
        if (nw < 0) {
            nz = -1;
            if (nw == -2) {
                nz = -2;
            }
            return nz;
        }
        return nz;
```

I've confirmed tests pass for me locally if this part is replaced with what follows the other call to `amos_mlri`.

ilayn replied:
This still doesn't explain why scipy 1.11.x can't match 1e-15 rtol bound you showed though.

steppi replied:
This is what I get running the example above on `main`

```
In [2]: run test_iv_miller.py
scipy: (-2.0645908843310795e-55-1.112777373104039e-55j)
mpmath: (-2.06459088433096e-55 - 1.11277737310399e-55j)
relative error: 5.64730905928374e-14
```

so not quite as good as this branch now, but still better than this branch before the fix. I suspect you probably improved something somewhere else when translating, and this is where the improvement comes from verse `main`.

ilayn commented:
I am not sure where I confused myself but I used my working code and swapped places following your lead and indeed tests pass. And moreover, I reverted the modified test since that also passes too. Let's see how it goes. Overall, I think this indeed was a very good catch. And the test suite is quite comprehensive it seems.

steppi reviewed (approved):
Thanks @ilayn. I've looked over everything else with a level of attention somewhere between spot check and exhaustive study while referring to the original paper. The rest looks good to me, and none of it has such twisted logic in the original as `binu`. It's possible I missed something still, but I feel confident now, especially with the level of testing we have.

ilayn commented:
Thanks @steppi much appreciated.

steppi commented:
> @steppi do you think we should update the docstrings as currently they state along the lines of: `` Wrapper for a C translation of the AMOS [1]_ routines `zairy` and `zbiry`. `` which I guess now is not 100% correct. I suggested in [ilayn#7](https://github.com/ilayn/scipy/pull/7) to change to `` Wrapper for a C translation of the AMOS [1]_ routines `zairy` and `zbiry`. `` however, perhaps this is too much detail for the end user. Do you have a view here?

Sorry. I didn't find any sign of https://github.com/ilayn/scipy/pull/7 mentioned in the comment chain above. I would have waited to merge this if I'd known you had this concern.

Honestly, I thought the docstrings were fine as is. I was thinking of the AMOS routines in the abstract, not tied to any particular implementation. Now that you mention it though, letting users know these are translations might save people from some confusion around our functions giving slightly different results from AMOS in some cases.

For the future. I expect these functions to diverge further and further from AMOS over time as we make improvements. If so,  I think the docstrings should be updated again to say our implementations are based on those of Amos, citing D.E. Amos' [Computation of Bessel functions of a complex argument](https://www.osti.gov/servlets/purl/6148453), which discusses the math behind the implementations.

ilayn commented:
> Sorry. I didn't find any sign of https://github.com/ilayn/scipy/pull/7 mentioned in the comment chain above. I would have waited to merge this if I'd known you had this concern. 

No problem at all. There is just too much to look at with these Fortran stuff so we are always forgetting stuff. 

But indeed I would also using the abstract terminology of citing Amos' paper. The interested readers would discover where we got the implementations at the top of the license comments anyways.

inkydragon commented on scipy/special/_amos.c:
```diff
@@ -0,0 +1,6217 @@
+        y[0] = 0.;
+        nz = 2;
+    }
+    if (n == 2) {
+        return nz;
+    }
+    fn = fnu + 1.;
+    ck = fn*rz;
```
Maybe there's a missing branch here?
Between L4009~L4010

```c
    if (nz == 0) {
        return nz;
    }
```

https://github.com/scipy/scipy/blob/e5f21298b62fb03732f000645d2caaa67aad8800/scipy/special/amos/zkscl.f#L52-L56

ilayn replied:
I think you are right. Very nice catch. Apologies for missing that. Would you like to send a PR?

inkydragon replied:
Yes, I'd be happy to submit a pr, however I'm having some difficulty constructing a test case.
Although the `amos_kscl` function affects almost all the `amos` exported functions, since it is not exported, it can only be tested through the upper level already exported functions.
Currently, the shortest call path seems to be:

```dot
amos_airy 
	-> amos_bknu -> amos_kscl

amos_besh, amos_besk
	-> amos_acon
	-> amos_bknu -> amos_kscl
```

I'm going to open a draft pr and then see what feedback I get from people.

ilayn replied:
Since this is a translation work, it is indeed very challenging to hit certain parts of the codebase unless called specifically internally but I think this case warrants a fix with or without a test case anyways, for obvious reasons. So no problem if you can't find a test case.
